src/components/content-tab-assistant.tsx

```diff
@@ -71,62 +77,153 @@ const OfflineModeView = () => {
 	);
 };
 
-const AuthenticatedView = memo(
-	( {
-		messages,
-		isAssistantThinking,
-		updateMessage,
-		siteId,
-	}: {
-		messages: MessageType[];
-		isAssistantThinking: boolean;
-		updateMessage: (
-			id: number,
-			codeBlockContent: string,
-			cliOutput: string,
-			cliStatus: 'success' | 'error',
-			cliTime: string
-		) => void;
-		siteId: string;
-	} ) => {
+type OnUpdateMessageType = (
+	id: number,
+	codeBlockContent: string,
+	cliOutput: string,
+	cliStatus: 'success' | 'error',
+	cliTime: string
+) => void;
+
+interface AuthenticatedViewProps {
+	messages: MessageType[];
+	isAssistantThinking: boolean;
+	updateMessage: OnUpdateMessageType;
+	siteId: string;
+}
+
+export const AuthenticatedView = memo(
+	( { messages, isAssistantThinking, updateMessage, siteId }: AuthenticatedViewProps ) => {
 		const endOfMessagesRef = useRef< HTMLDivElement >( null );
+		const [ showThinking, setShowThinking ] = useState( false );
+		const lastMessage = useMemo(
+			() =>
+				showThinking
+					? ( { role: 'assistant', id: -1, createdAt: Date.now() } as MessageType )
+					: messages[ messages.length - 1 ],
+			[ messages, showThinking ]
+		);
+		const messagesToRender =
+			messages[ messages.length - 1 ]?.role === 'assistant'
+				? messages.slice( 0, -1 )
+				: [ ...messages ];
+		const showLastMessage = useMemo(
+			() => showThinking || messages[ messages.length - 1 ]?.role === 'assistant',
+			[ messages, showThinking ]
+		);
 
 		useEffect( () => {
 			const timer = setTimeout( () => {
 				if ( endOfMessagesRef.current ) {
 					endOfMessagesRef.current.scrollIntoView( { behavior: 'smooth' } );
 				}
-			}, 100 ); // Slight delay to ensure DOM updates
+			}, 400 );
+			return () => clearTimeout( timer );
+		}, [ messages?.length, isAssistantThinking, showThinking, showLastMessage ] );
 
+		useEffect( () => {
+			let timer: NodeJS.Timeout;
+			if ( isAssistantThinking ) {
+				timer = setTimeout( () => setShowThinking( true ), MIMIC_CONVERSATION_DELAY );
+			} else {
+				setShowThinking( false );
+			}
 			return () => clearTimeout( timer );
-		}, [ messages?.length ] );
+		}, [ isAssistantThinking ] );
 
-		if ( messages.length === 0 ) {
-			return null;
-		}
+		const RenderMessage = useCallback(
```

Comment:
I wonder if memoizing this and the `RenderLastMessage` function components will have an impact on performance. I often use `useCallback` for handler functions passed as props to components in order to avoid re-renders, but not sure if in this is the same case.

Reply:
React components are actually function, so probably is the same pretty much.
I can try `React.memo` however if we think this makes more sense

Reply:
I tried to switch, but the animations became sluggish. I'll leave it as is.